crates/iceberg/src/arrow.rs

```diff
@@ -106,3 +114,224 @@ impl ArrowReader {
+        if let Some(doc) = &field.doc {
+            metadata.insert(DOC.to_string(), doc.clone());
+        }
+        Ok(ArrowSchemaOrFieldOrType::Field(
+            ArrowField::new(field.name.clone(), ty, !field.required)
+                .with_metadata(metadata)
+                .into(),
+        ))
+    }
+
+    fn r#struct(
+        &mut self,
+        _: &crate::spec::StructType,
+        results: Vec<Self::T>,
+    ) -> crate::Result<Self::T> {
+        let fields = results
+            .into_iter()
+            .map(|result| match result {
+                ArrowSchemaOrFieldOrType::Field(field) => field,
+                _ => unreachable!(),
+            })
+            .collect();
+        Ok(ArrowSchemaOrFieldOrType::Type(ArrowType::Struct(fields)))
+    }
+
+    fn list(&mut self, list: &crate::spec::ListType, value: Self::T) -> crate::Result<Self::T> {
+        let field = match self.field(&list.element_field, value)? {
+            ArrowSchemaOrFieldOrType::Field(field) => field,
+            _ => unreachable!(),
+        };
+        Ok(ArrowSchemaOrFieldOrType::Type(ArrowType::List(field)))
+    }
+
+    fn map(
+        &mut self,
+        map: &crate::spec::MapType,
+        key_value: Self::T,
+        value: Self::T,
+    ) -> crate::Result<Self::T> {
+        let key_field = match self.field(&map.key_field, key_value)? {
+            ArrowSchemaOrFieldOrType::Field(field) => field,
+            _ => unreachable!(),
+        };
+        let value_field = match self.field(&map.value_field, value)? {
+            ArrowSchemaOrFieldOrType::Field(field) => field,
+            _ => unreachable!(),
+        };
+        let field = ArrowField::new(
+            "entries",
+            ArrowType::Struct(vec![key_field, value_field].into()),
+            map.value_field.required,
+        );
+
+        Ok(ArrowSchemaOrFieldOrType::Type(ArrowType::Map(
+            field.into(),
+            false,
+        )))
+    }
+
+    fn primitive(&mut self, p: &crate::spec::PrimitiveType) -> crate::Result<Self::T> {
+        match p {
+            crate::spec::PrimitiveType::Boolean => {
+                Ok(ArrowSchemaOrFieldOrType::Type(ArrowType::Boolean))
+            }
+            crate::spec::PrimitiveType::Int => Ok(ArrowSchemaOrFieldOrType::Type(ArrowType::Int32)),
+            crate::spec::PrimitiveType::Long => {
+                Ok(ArrowSchemaOrFieldOrType::Type(ArrowType::Int64))
+            }
+            crate::spec::PrimitiveType::Float => {
+                Ok(ArrowSchemaOrFieldOrType::Type(ArrowType::Float32))
+            }
+            crate::spec::PrimitiveType::Double => {
+                Ok(ArrowSchemaOrFieldOrType::Type(ArrowType::Float64))
+            }
+            crate::spec::PrimitiveType::Decimal { precision, scale } => {
+                Ok(ArrowSchemaOrFieldOrType::Type(ArrowType::Decimal128(
+                    TryInto::try_into(*precision).map_err(|err| {
```

Comment:
How about using [this method](https://docs.rs/arrow/latest/arrow/datatypes/fn.validate_decimal_precision_and_scale.html) to do the check?